### ISSUE TITLE
Notebooks using DiscreteDP type

### ISSUE THREAD

**Issue by MaximilianJHuber** [QuantEcon/QuantEcon.jl#113]
Hi,

I fell in love with Julia only last week and started learning it due to performance reasons when doing discrete dynamic programming (I used to code in Matlab and R).

So far I am missing notebooks using the discreteDP type and appropriate functions.
If you think it is a good idea to port the Python notebook [Discrete Dynamic Programming](http://quant-econ.net/py/discrete_dp.html) and maybe later [The Aiyagari Model](http://quant-econ.net/py/aiyagari.html) or to write an bigger example where Modified Policy Function Iteration makes sense, I can get right on it.

Max

**Comment by spencerlyon2** [comment]
HI @MaximilianJHuber

Contributions of this kind are very welcome.

I see that you have closed the issue. If you are still willing to work on it let us know and we can try to help you get started.

Thanks again

**Comment by spencerlyon2** [comment]
HI @MaximilianJHuber

Contributions of this kind are very welcome.

I see that you have closed the issue. If you are still willing to work on it let us know and we can try to help you get started.

Thanks again

**Comment by MaximilianJHuber** [comment]
Sure! I closed because issue 83 has this covered.

**Comment by spencerlyon2** [comment]
Ahh, I see. Thanks for doing that.

Given that #83 is more of a collection of things, rather than discussion for specific topics I think I'll reopen this issue and make a note over in #83 that we are working on it here.

Do you already have an idea of what you would like to do to get started? If so, please go ahead and start when you can and make sure to ask if you have any questions. If you would like more guidance on what to do please don't hesitate to ask. We're happy to help you get going.

Thanks again!

**Comment by spencerlyon2** [comment]
Hey @MaximilianJHuber  thanks for doing that.

Due to copyright restrictions the source for the lectures must remain private. Once the updates you have been working on are finished and merged into the appropriate repositories we will make sure to update the lecture also.  (cc @jstac)

Looking forward to seeing your code!

**Comment by MaximilianJHuber** [comment]
I see. You can find the code for the example in the lecture [here](http://nbviewer.jupyter.org/github/MaximilianJHuber/DDP/blob/master/discrete-dp-v3.ipynb). I would be very happy about comments!

Also, I coded the [first part](http://nbviewer.jupyter.org/github/MaximilianJHuber/DDP/blob/master/discrete-dp-ex.ipynb) of the exercise. Either JuliaBox performance is bad, my code is bad, or there is something strange going on in ddp.jl, but when I call the DiscreteDP-constructor with a grid_size=1500 the kernel goes bye-bye. Running the notebook with grid_size=500 is fine. A performance comparison with the Python [notebook](http://nbviewer.jupyter.org/github/QuantEcon/QuantEcon.applications/blob/master/discrete_dp/discrete_dp_solutions_py.ipynb) would be interesting, but apparently I can't even correctly install numba.

**Comment by oyamad** [comment]
@MaximilianJHuber Thanks for your work!

I ran your notebook, and the constructor did work in my environment, which took 15 seconds. And `solve(ddp, PFI)` took more than 10 minutes (I tried it only once). You code does not seem to have any problem. I am afraid the ddp.jl code has not been optimized.

As for the Python version, did you try installing Anaconda? (You might look at [this page](http://quant-econ.net/py/getting_started.html) for instruction.) It includes Numba.

**Comment by MaximilianJHuber** [comment]
I managed to install Anaconda and Julia locally (2.66GHZ, 4GB RAM) and ran both notebooks with **grid_size=500**. The Julia version took repeatedly about 11s, the Python version about 50ms.
I opend a new issue  #118.
I will read Sengupta's "Julia High Performance" first, maybe I can then contribute to this problem.

Anyway, I will finish up the solution to the exercise in the next days.

**Comment by spencerlyon2** [comment]
I haven't checked so I can't be sure, but I would imagine the problem is related to type instability at an "outer" level of the algorithm that causes all inner routines to be inefficient. 

One would check by trying `@code_warntype func(args...)` for all `func` that is called when you use one of these DiscreteDP solution routines.

**Comment by MaximilianJHuber** [comment]
@spencerlyon2 definitly a good idea! Meanwhile, I proposed a possible improvement in issue #118.

Both the [example](http://nbviewer.jupyter.org/github/MaximilianJHuber/DDP/blob/master/discrete-dp-v3.ipynb) and the [exercise](http://nbviewer.jupyter.org/github/MaximilianJHuber/DDP/blob/master/discrete-dp-ex.ipynb) related to the Python lecture [Discrete Dynamic Programming](http://quant-econ.net/py/discrete_dp.html) are finished from my side.

**Comment by MaximilianJHuber** [comment]
I finished the [Julia code](http://nbviewer.jupyter.org/github/MaximilianJHuber/DDP/blob/master/aiyagari_household.ipynb) for the Python lecture on [the Aiyagari Model](http://quant-econ.net/py/aiyagari.html) ready. I would appreciate if someone could have a look at the Houldhold-type and related functions. I could not easily translate the class structure from Python, because Julia is not that object oriented.

**Comment by spencerlyon2** [comment]
@MaximilianJHuber that's great, thanks for working on it.

Are you comfortable opening a pull request in [QuantEcon.applications](https://github.com/QuantEcon/QuantEcon.applications) that includes your work so we can review it?
